src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,459 @@
+
+  math::ClearAlias(dataset);
+  math::ClearAlias(labels);
+
+  dataset = std::move(newDataset);
+  labels = std::move(newLabels);
+
+  // Re-calculate target neighbors as indices changed.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+}
+
+inline void Projection(arma::mat& iterate)
+{
+  // Projection.
+  arma::vec eigval;
+  arma::mat eigvec;
+
+  arma::eig_sym(eigval, eigvec, iterate);
+
+  arma::uvec ind = arma::find(eigval > 0);
+  arma::mat diagEigVal = arma::diagmat(eigval);
+  iterate = eigvec.cols(ind) * diagEigVal.submat(ind, ind) *
+      arma::trans(eigvec.cols(ind));
+}
+
+//! Evaluate cost over whole dataset.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates)
+{
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_elem)
+    Projection(iterate);
+
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  // Re-calculate impostors on transformed dataset.
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (int j = k - 1; j >= 0; j--)
+    {
+      // Bound constraints to avoid uneccesary computation.
+      for (size_t l = 0, bp = k; l < bp ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+        double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i))) -
+                      metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(impostors(l, i)));
+
+        // Check bounding condition.
+        if (eval < -1)
+        {
+          // update bound.
+          bp = l;
+          break;
+        }
+
+        cost += regularization * (1 + eval);
+      }
+    }
+  }
+
+  return cost;
+};
+
+//! Calculate cost over batches.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
+                                          const size_t begin,
+                                          const size_t batchSize)
+{
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_elem)
+    Projection(iterate);
+
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
```

Comment:
I didn't realize it when I reviewed, but it looks like every use of `coordinates` implies that the implementation here actually is optimizing `L` for `M = L^T L`.  That means two things:

 * No projection step is necessary---for any `L`, `M` is guaranteed to be positive semidefinite.

 * It should be possible to allow the user to specify low-rank `L`: that is, instead of being `d x d`, `L` can have size `r x d` for some user-specified `r <= d`.  Depending on the dataset, some smaller values of `r` may give degraded performance.  But, small `r` will generally both accelerate the KNN search in the transformed space since it is lower-dimensional, and also accelerate the optimization process since the matrices are smaller.

In general for your optimization efforts I would say we should add this but don't get hung up if it is not giving perfectly good performance.

Reply:
>  it looks like every use of coordinates implies that the implementation here actually is optimizing L for M = L^T L.

Can you please explain some more, I can't seem to get it.

Reply:
So, in the paper, we are trying to find some Mahalanobis distance matrix `M`.  This matrix `M` improves classification performance when we use the distance function `d_M(x, y) = (x - y)^T M (x - y)`.

But note that if we stretch the dataset (as is done in the code) by some matrix `L` as `X' = L*X`, this would give the distance function `d'(x, y) = (Lx - Ly)^T (Lx - Ly) = (x - y)^T L^T L (x - y)`.  Hence, `L` must be such that `M = L^T L`.

So, if we want to optimize over `M`, then the code should be, e.g., `transformedDataset = arma::chol(coordinates) * dataset` (another decomposition could work too, maybe LU is faster?).  But in the end, I'd like to optimize over `L`, because then we can also make it low-rank which could give additional speedup (and was the original idea behind LMNN+LRSDP anyway).

Does that make sense?  Let me know if I can clarify further.

Reply:
> if we stretch the dataset (as is done in the code) by some matrix L as X' = L*X, this would give the distance function d'(x, y) = (Lx - Ly)^T (Lx - Ly) = (x - y)^T L^T L (x - y). Hence, L must be such that M = L^T L.

Thanks, this helps a lot. I was missing the point that we are actually learning L here, it became clear as you explained.